voter-dapp/src/ActiveRequests.js

```diff
@@ -49,8 +51,87 @@ function ActiveRequests() {
     // Each of the subfetches has to complete. In drizzle, `undefined` means incomplete, while `null` means complete
     // but the fetched value was null, e.g., no `comittedValue` existed.
     voteStatuses.every(voteStatus => voteStatus.committedValue !== undefined && voteStatus.hasRevealed !== undefined);
+  // Future hook calls that depend on `voteStatuses` should use `voteStatusesStringified` in their dependencies array
+  // because `voteStatuses` will never compare equal after a re-render, even if no values in it actually changed.
+  // Also note that `JSON.stringify` doesn't distinguish `undefined` and `null`, but in Drizzle, those mean different
+  // things and should retrigger dependent hooks.
+  const voteStatusesStringified = JSON.stringify(voteStatuses, (k, v) => (v === undefined ? "undefined" : v));
 
-  if (!initialFetchComplete || !subsequentFetchComplete || !revealEvents) {
+  // The decryption key is a function of the account and `currentRoundId`, so we need the user to re-sign a message
+  // any time one of those changes. We also don't want to show multiple, identical notifications when this effect gets
+  // cleaned up on unrelated re-renders. The best solution I've come up with so far is to store the keys in a nested map
+  // (i.e., Object) from account => roundId => key. When a signing request is sent out (i.e., the user is shown a
+  // Metamask popup), we inject a special `messageSigningProcessingToken` at `decryptionKeys[account][currentRoundId]`
+  // to prevent resending that request. Note that in this approach, we don't disregard the result of this hook on
+  // component unmount.
+  const [decryptionKeys, setDecryptionKeys] = useState({});
+  const messageSigningProcessingToken = "processing";
+  useEffect(() => {
+    async function getDecryptionKey() {
+      if (!account || !currentRoundId) {
+        return;
+      }
+      if (decryptionKeys[account] && decryptionKeys[account][currentRoundId]) {
+        return;
+      }
+      setDecryptionKeys(prev => ({
+        ...prev,
+        [account]: { ...prev[account], [currentRoundId]: messageSigningProcessingToken }
+      }));
+      // TODO(ptare): Handle the user refusing to sign the message.
+      const { privateKey } = await deriveKeyPairFromSignatureMetamask(web3, getKeyGenMessage(currentRoundId), account);
+      setDecryptionKeys(prev => ({ ...prev, [account]: { ...prev[account], [currentRoundId]: privateKey } }));
+    }
+
+    getDecryptionKey();
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [account, currentRoundId, decryptionKeys]);
```

Comment:
As discussed IRL, is `decryptionKeys` required here?

Also, if this memoization works correctly (no spurious calls), could we completely remove the `messageSigningProcessingToken` stage of this?

Reply:
Done.